src/GenAPI/Microsoft.DotNet.GenAPI/IMethodSymbolExtensions.cs

```diff
@@ -0,0 +1,99 @@
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.DotNet.ApiSymbolExtensions.Filtering;
+using Microsoft.DotNet.ApiSymbolExtensions;
+
+namespace Microsoft.DotNet.GenAPI
+{
+    internal static class IMethodSymbolExtensions
+    {
+        /// <summary>
+        /// Compare the equality of two method signatures for the purpose of emitting a "new"
+        /// keyword on a method's return type. This is *not* meant to be complete implementation,
+        /// but rather a heuristic to check that one method may hide another.
+        /// </summary>
+        /// <param name="method"></param>
+        /// <param name="otherMethod"></param>
+        /// <returns>true if the signatures are determined to be equal</returns>
+        public static bool SignatureEquals(this IMethodSymbol? method, IMethodSymbol? otherMethod)
+        {
+            if (method is null || otherMethod is null)
+            {
+                return false;
+            }
+
+            if (method.Equals(otherMethod, SymbolEqualityComparer.Default))
+            {
+                return true;
+            }
+
+            if (method.Name != otherMethod.Name)
+            {
+                return false;
+            }
+
+            if (method.Arity != otherMethod.Arity || method.Parameters.Length != otherMethod.Parameters.Length)
+            {
+                return false;
+            }
+
+            // compare parameter types
+            for (int i = 0; i < method.Parameters.Length; i++)
+            {
+                if (!method.Parameters[i].Type.Equals(otherMethod.Parameters[i].Type, SymbolEqualityComparer.Default))
+                {
+                    return false;
+                }
+            }
+
+            // TODO: GenAPI does not currently preserve __arglist as a parameter.
+            // Add test case for this branch when that is fixed.
+            if (method.IsVararg != otherMethod.IsVararg)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        public static bool IsImplicitDefaultConstructor(this IMethodSymbol method, ISymbolFilter symbolFilter)
+        {
+            // ensure this is a parameterless constructor
+            if (method.MethodKind != MethodKind.Constructor || !method.Parameters.IsEmpty)
+            {
+                return false;
+            }
+
+            // see https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/classes#15115-default-constructors
+            // these are the only two accessibilities given to the implicit default constructor, regardless of containing type accessibility.
+            if (method.ContainingType.IsAbstract ?
+                    method.DeclaredAccessibility != Accessibility.Protected :
+                    method.DeclaredAccessibility != Accessibility.Public)
```

Comment:
nit: indentation doesn't look right.

Reply:
I intentionally indented here to show these were the two operands to the ternary operator above.  It made it more readable to me (and Format Document didn't correct it), but I'm fine to remove it.